### PR TITLE
Shield Fixed Key Encryption

### PR DESCRIPTION
**Pull request by daviddob**
- Introduces a new boolean for jobs, allowing operators to mark them as "DR" jobs that use a fixed key.
- Introduces a Disaster Key used to derive fixed encryption keys used for fixed-key jobs provided at initialization time.
- On shield rekey, operators will be given the option to rotate their Disaster Key as well; in which the same process from shield init is followed.

### CODE REVIEW

File: client/v2/shield/init.go [hunk 1]
```diff
@@ -1,10 +1,32 @@
...
 		Master string `json:"master"`
 	}{
 		Master: master,
 	}
-	return c.post("/v2/init", in, nil)
+
+	out := struct {
+		DisasterKey string `json:"disaster_key"`
+	}{
+		DisasterKey: "",
```
Comment (jhunt) [inline client/v2/shield/init.go]: You don't need to initialize this anon structure var.

File: client/v2/shield/init.go [hunk 2]
```diff
@@ -1,10 +1,32 @@
...
+
+func (c *Client) SplitKey(s string, n int) string {
+	var buffer bytes.Buffer
+	for i, rune := range s {
+		buffer.WriteRune(rune)
+		if i%n == (n-1) && i != (len(s)-1) {
+			buffer.WriteRune('\n')
+		}
+	}
+	return buffer.String()
```
Comment (jhunt) [inline client/v2/shield/init.go]: There's a more legible way of doing this (breaking a long string into lines) using string slices; we should investigate.

File: client/v2/shield/rekey.go [hunk 3]
```diff
@@ -1,12 +1,23 @@
...
 		OldMaster: oldmaster,
 		NewMaster: newmaster,
+		RekeyDR:   rekeyDR,
 	}
-	return c.post("/v2/rekey", in, nil)
+
+	out := struct {
+		DisasterKey string `json:"disaster_key"`
+	}{
+		DisasterKey: "",
```
Comment (jhunt) [inline client/v2/shield/rekey.go]: You don't need to initialize this anon structure var.

File: cmd/buckler/main.go [hunk 4]
```diff
@@ -83,6 +83,7 @@ var opts struct {
 	Rekey struct {
 		OldMaster string `cli:"--old-master"`
 		NewMaster string `cli:"--new-master"`
+		RekeyDR   bool   `cli:"--rekey-dr"`
```
Comment (jhunt) [inline cmd/buckler/main.go]: Can we rename this switch to `--rotate-disaster-key`?

File: cmd/buckler/main.go [hunk 5]
```diff
@@ -842,9 +855,19 @@ func main() {
...
+		disasterKey, err := c.Rekey(opts.Rekey.OldMaster, opts.Rekey.NewMaster, opts.Rekey.RekeyDR)
 		bail(err)
 
+		if disasterKey != "" {
+			fmt.Printf("@R{BELOW IS YOUR DISASTER KEY FOR RECOVERING FIXED-KEY BACKUPS.}\n")
+			fmt.Printf("@R{SAVE THIS IN A SECURE LOCATION.}\n")
+			fmt.Printf("----------------------------------------------------------------\n")
+			fmt.Printf("@Y{" + c.SplitKey(disasterKey, 64) + "}")
+			fmt.Printf("\n----------------------------------------------------------------\n")
+		} else if opts.Rekey.RekeyDR {
```
Comment (jhunt) [inline cmd/buckler/main.go]: Is it possible for Rekey() to return a nil err and an empty disaster key?  If it is, we should fix that.

File: cmd/buckler/main.go [hunk 6]
```diff
@@ -812,11 +815,21 @@ func main() {
...
 
 		fmt.Printf("SHIELD core unlocked successfully.\n")
 
+		if disasterKey != "" {
+			fmt.Printf("@R{BELOW IS YOUR DISASTER KEY FOR RECOVERING FIXED-KEY BACKUPS.}\n")
+			fmt.Printf("@R{SAVE THIS IN A SECURE LOCATION.}\n")
+			fmt.Printf("----------------------------------------------------------------\n")
+			fmt.Printf("@Y{" + c.SplitKey(disasterKey, 64) + "}")
+			fmt.Printf("\n----------------------------------------------------------------\n")
+		} else {
```
Comment (jhunt) [inline cmd/buckler/main.go]: Is it possible for Initialize() to return a nil err and an empty disaster key? If it is, we should fix that.

File: cmd/buckler/main.go [hunk 7]
```diff
@@ -2103,9 +2126,9 @@ func main() {
 			)
 		*/
 		/* FIXME: support --long / -l and maybe --output / -o "fmt-str" */
-		tbl := tui.NewTable("UUID", "Name", "Summary", "Schedule", "Status", "Policy", "SHIELD Agent", "Target", "Store")
+		tbl := tui.NewTable("UUID", "Name", "Summary", "Schedule", "Status", "Policy", "SHIELD Agent", "Target", "Store", "Disaster Recovery")
 		for _, job := range jobs {
-			tbl.Row(job, job.UUID, job.Name, job.Summary, job.Schedule, job.Status(), job.Policy.Name, job.Agent, job.Target.Name, job.Store.Name)
+			tbl.Row(job, job.UUID, job.Name, job.Summary, job.Schedule, job.Status(), job.Policy.Name, job.Agent, job.Target.Name, job.Store.Name, job.DisasterRecovery)
```
Comment (jhunt) [inline cmd/buckler/main.go]: What do you think about putting a "Type" column before "Name", and setting it to "normal" if not a DR job, and "DR" if it is (probably with some coloring for boldness), vs. having a dedicated "Disaster Recovery" column?

File: cmd/buckler/main.go [hunk 8]
```diff
@@ -2147,6 +2170,7 @@ func main() {
 		r.Add("Storage Plugin", job.Store.Plugin)
 		r.Break()
 
+		r.Add("Disaster Recovery", strconv.FormatBool(job.DisasterRecovery))
```
Comment (jhunt) [inline cmd/buckler/main.go]: If we do the "Type" thing in the table, we should do the same here.

File: db/schema_v4.go [hunk 9]
```diff
@@ -87,17 +87,18 @@ func (s v4Schema) Deploy(db *DB) error {
...
+	               store_uuid         UUID NOT NULL,
+	               tenant_uuid        UUID,
+	               schedule           TEXT NOT NULL,
+	               next_run           INTEGER DEFAULT 0,
+	               retention_uuid     UUID NOT NULL,
+	               priority           INTEGER DEFAULT 50,
+	               paused             BOOLEAN,
+	               name               TEXT,
+				   summary            TEXT,
+				   disaster_recovery  INTEGER DEFAULT 0
```
Comment (jhunt) [inline db/schema_v4.go]: You've got some whitespace issues here...

File: web2/htdocs/index.html [hunk 10]
```diff
@@ -243,6 +243,36 @@ <h2>CLI Authentication Credentials</h2>
...
           </div>
     <!-- }}} --></script>
+    <script type="text/html" id="tpl--disasterkey"><!-- {{{ -->
+        [[#
+          {disasterkey} template
+  
+          Displays the Shield Disaster Key, so that operators can restore
+          fixed-key backups in the event of a disaster.
+        ]]
+            <div id="cliauth" class="frontdoor">
```
Comment (jhunt) [inline web2/htdocs/index.html]: This shouldn't be "cliauth" - probably need to rework CSS if it doesn't fall under `.frontdoor`

File: web2/htdocs/index.html [hunk 11]
```diff
@@ -243,6 +243,36 @@ <h2>CLI Authentication Credentials</h2>
...
+              <div>
+                <h1>SHIELD<span>Production<span></h1>
```
Comment (jhunt) [inline web2/htdocs/index.html]: Shouldn't `Production` come from our SHIELD env?

File: web2/htdocs/index.html [hunk 12]
```diff
@@ -243,6 +243,36 @@ <h2>CLI Authentication Credentials</h2>
...
+                  <p>Here is your SHIELD Disaster Key:</p>
+  
+                  <pre id="disasterkey">[[= _.disaster_key ]]</pre>
+  
+                  <p id="dr-explain">
+                    You will need to save this key in a secure location
+                    for use in recovery of fixed-key backups in a disaster scenario.
+                    Fixed-key backups must be used to backup shield itself. Once you
+                    leave this screen, there is no way to access this key or corresponding
+                    fixed-key backups unless you have saved it!
```
Comment (jhunt) [inline web2/htdocs/index.html]: We should either change all the other "Disaster Recovery" references to "Fixed Key", or change these "Fixed Key" references to "Disaster Recovery", for consistency.

File: web2/htdocs/index.html [hunk 13]
```diff
@@ -243,6 +243,36 @@ <h2>CLI Authentication Credentials</h2>
...
+                  </p>
+                  <a href="/"><button id="accept">I Understand</button></a>
```
Comment (jhunt) [inline web2/htdocs/index.html]: What does the accept button do?

File: web2/htdocs/index.html [hunk 14]
```diff
@@ -1029,6 +1059,15 @@ <h3>Configure a New Protected System</h3>[[# {{{ ]]
             <div id="choose-plugin"></div>
             <div id="configure-plugin"></div>
 
+            <div class="ctl" data-field="fixed-key">
+                <label for="target-fixed-key">Fixed-Key Encryption?</label>
```
Comment (jhunt) [inline web2/htdocs/index.html]: Make the terminology consistent.

File: web2/htdocs/shield.css [hunk 15]
```diff
@@ -865,7 +865,7 @@ form .ctl label {
   line-height: 20px;
   display: block;
 
-  width: 40%;
+  width: 50%;
```
Comment (jhunt) [inline web2/htdocs/shield.css]: How well does this play with the error reporting on form validation?

File: cmd/buckler/main.go (hunk 6 above)
Comment (daviddob) [reply cmd/buckler/main.go]: It should not be possible (at least from the API standpoint always a bad, fail, or they key will exist) these checks were more to ensure future api changes dont introduce a bug that isnt caught quickly.

File: cmd/buckler/main.go (hunk 5 above)
Comment (daviddob) [reply cmd/buckler/main.go]: (same as above)

File: cmd/buckler/main.go (hunk 7 above)
Comment (daviddob) [reply cmd/buckler/main.go]: Type would make a bit more sense if we had more than two and if we expected operators to use fixed-key backups for more than just shield or a rare other case.

File: web2/htdocs/index.html (hunk 13 above)
Comment (daviddob) [reply web2/htdocs/index.html]: Just redirects to the landing page which happens to be same as /systems

File: web2/htdocs/shield.css (hunk 15 above)
Comment (daviddob) [reply web2/htdocs/shield.css]: Seems to work okay, errors still appear as they used to

**Pull request merged**